torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2032,18 +2038,39 @@ def _get_grad_norm(
         raise ValueError(
             f"Requires uniform dtype across all gradients but got {grad_dtypes}"
         )
-    # Compute the gradient norm in FP32, where we treat the gradients as a
-    # single vector
+
+    first_device = grads[0].device
+    grouped_grads: Dict[
+        Tuple[torch.device, torch.dtype], List[List[torch.Tensor]]
+    ] = _group_tensors_by_device_and_dtype(
+        [[g.detach() for g in grads]]
+    )  # type: ignore[assignment]
+
+    # Compute the gradient norm in FP32, where we treat the gradients as a single vector
+    grad_norms = []
+    for (device, _), [grads] in grouped_grads.items():
+        if (foreach is None or foreach) and _has_foreach_support(grads, device=device):
+            grad_norms.extend(
+                torch._foreach_norm(grads, norm_type, dtype=torch.float32)
+            )
+        elif foreach:
+            raise RuntimeError(
+                f"foreach=True was passed, but can't use the foreach API on {device.type} tensors"
+            )
+        else:
+            grad_norms.extend(
+                [
+                    torch.linalg.vector_norm(g, norm_type, dtype=torch.float32)
+                    for g in grads
+                ]
+            )
+
     grad_norm = torch.linalg.vector_norm(
```

Comment:
Wait...isn't this vector_norm call redundant after the for loop?

Reply:
Not sure what you mean, you run vector_norm once per param and then aggregate the results for all params by running vector_norm on the list of norms.

Reply:
Ah, sorry for the misunderstanding

Reply:
No worries